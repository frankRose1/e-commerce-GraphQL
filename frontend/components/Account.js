import React, { Component } from 'react';
import gql from 'graphql-tag';
import {Query} from 'react-apollo';
import {formatDistance} from 'date-fns';
import styled from 'styled-components';
import Link from 'next/link';
import formatMoney from '../lib/formatMoney';
import OrderItem from './styles/OrderItemStyles';
import Error from './ErrorMessage';
import Loading from './Loading';
import SubTitle from './styles/SubTitleStyles';

const OrdersGrid = styled.ul`
  display: grid;
  grid-gap: 4rem;
  grid-template-columns: repeat(auto-fit, minmax(40%, 1fr));
`;

const USER_ORDERS_QUERY = gql`
  query USER_ORDERS_QUERY {
    userOrders(orderBy: createdAt_DESC) {
      id
      total
      createdAt
      items {
        id
        quantity
        description
        title
        price
        image
      }
    }
  }
`;

class Account extends Component {
  render() {
    return (
      <Query query={USER_ORDERS_QUERY}>
        {({data, loading, error}) => {
          const {userOrders} = data;
          if (loading) return <Loading />
          if(error) return <Error error={error} />
          if (!userOrders.length) return <p>Looks like you havn't ordered anything yet!</p>
          
          return (
            <div>
              <SubTitle>Your Account</SubTitle>
              <h2>You have {userOrders.length} order{userOrders.length === 1 ? '' : 's'}</h2>
              <OrdersGrid>
                {userOrders.map(order => (
                  <OrderItem key={order.id}>
                    <Link href={{
                      pathname: '/order',
                      query: { orderId: order.id }
                    }}>
                      <a>
                        <div className="order-meta">
                          <p>{order.items.reduce((a, b) => a + b.quantity, 0)} Items</p>
                          <p>{order.items.length} Product{order.items.length === 1 ? '' : 's'}</p>
                          <p>{formatDistance(order.createdAt, new Date())}</p>
                          <p>{formatMoney(order.total)}</p>
                        </div>
                        <div className="images">
                          {order.items.map(item => (
                            <img 
                              src={item.image} 
                              alt={item.title}
                              key={item.id} />
                          ))}
                        </div>
                      </a>
                    </Link>
                  </OrderItem>
                ))}
            </OrdersGrid>
            </div>
          )
        }}
      </Query>
    );
  }
}

export default Account;