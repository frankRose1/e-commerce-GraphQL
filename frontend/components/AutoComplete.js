import React, { Component } from 'react';
import {ApolloConsumer} from 'react-apollo'; //run queries on demand
import gql from 'graphql-tag';
import Router from 'next/router';
import debounce from 'lodash.debounce';
import Downshift, {resetIdCounter} from 'downshift';
import {DropDown, DropDownItem, SearchStyles} from './styles/DropDown';

function routeToItem(item){
  Router.push({
    pathname: '/item',
    query: { id: item.id }
  })
}

//serach by either the title or description
const SEARCH_ITEMS_QUERY = gql`
  query SEARCH_ITEMS_QUERY($searchTerm: String!){
    items(where: {
      OR: [
        { title_contains: $searchTerm},
        { description_contains: $searchTerm}
      ]
    }){
      id
      image
      title
    }
  }
`;

class AutoComplete extends Component {

  state= {
    loading: false,
    items: []
  }

  //Manually fire off a query on input change
  handleChange = debounce(async (e, {query}) => {
    this.setState({loading: true});
    const res = await query({
      query: SEARCH_ITEMS_QUERY,
      variables: { searchTerm: e.target.value }
    });
    this.setState({
      loading: false,
      items: res.data.items
    })
  }, 400);

  render() {
    const {items, loading} = this.state;
    resetIdCounter();
    return (
      <SearchStyles>
        <Downshift
          onChange={routeToItem}
          itemToString={item => (item === null ? '' : item.title)}>
          { ({getInputProps, getItemProps, isOpen, inputValue, highlightedIndex}) => (
          <div>
            <ApolloConsumer>
              { client => (
                <input
                  {...getInputProps({
                    placeholder: "Search Our Awesome Collection",
                    type: "search",
                    id: 'search',
                    className: this.state.loading ? 'loading' : '',
                    onChange: e => {
                      e.persist();
                      this.handleChange(e, client);
                    }
                  })} />
              )}
            </ApolloConsumer>
            {isOpen && (
              <DropDown>
                {items.map((item, i) => (
                  <DropDownItem
                    {...getItemProps({item})}
                    highlighted={highlightedIndex === i}
                    key={item.id}>
                    <img width="50" src={item.image} alt={item.title}/>
                    {item.title}
                  </DropDownItem>
                ))}
                {!items.length && !loading && (
                  <DropDownItem>
                    Nothing found for "{inputValue}"
                  </DropDownItem>
                )}
              </DropDown>
            )}
          </div>
          )}
        </Downshift>
      </SearchStyles>
    );
  }
}

export default AutoComplete;