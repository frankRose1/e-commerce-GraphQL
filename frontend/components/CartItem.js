import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import formatMoney from '../lib/formatMoney';
import RemoveFromCart from './RemoveFromCart';

const CartItemStyles = styled.li`
  padding: 1rem 0;
  display: grid;
  border-bottom: 1px solid ${props => props.theme.lightgrey};
  grid-template-columns: auto 1fr auto;
  align-items: center;
  img {
    width: 100px;
    margin-right: 10px;
  }
  h3, p {
    margin: 0;
  }
`;

const DeletedItem = styled.li`
  padding: 1rem 0;
  display: grid;
  border-bottom: 1px solid ${props => props.theme.lightgrey};
  grid-template-columns: 1fr auto;
  align-items: center;
`;

const CartItem = ({cartItem: { item, quantity, id} }) => {
  //in the case that An item is deleted in the DB we dont want to show it in the users cart
  if (!item){
    return (
      <DeletedItem>
        <p>Sorry, this item is no longer sold in our store!</p>
        <RemoveFromCart cartItemId={id}/>
      </DeletedItem>
    )
  }

  return (
  <CartItemStyles>
    <img src={item.image} alt={item.title}/>
    <div className="cart-item-details">
      <h3>{item.title}</h3>
      <p>
        {formatMoney(item.price * quantity)}
        {' - '}
        <em>{quantity} &times; {formatMoney(item.price)} each</em>
      </p>
    </div>
    <RemoveFromCart cartItemId={id}/>
  </CartItemStyles>
)};

CartItem.propTypes = {
  cartItem: PropTypes.shape({
    quantity: PropTypes.number.isRequired,
    id: PropTypes.string.isRequired,
    item: PropTypes.object
  }).isRequired
};

export default CartItem;