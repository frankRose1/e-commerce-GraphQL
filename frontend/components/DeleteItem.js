import React, { Component } from 'react';
import {Mutation} from 'react-apollo';
import gql from 'graphql-tag';
import PropTypes from 'prop-types';
import {ALL_ITEMS_QUERY} from './Items';

const DELETE_ITEM_MUTATION = gql`
  mutation DELETE_ITEM_MUTATION($id: ID!){
    deleteItem(id: $id){
      id
    }
  }
`;

class DeleteItem extends Component {

  static propTypes = {
    id: PropTypes.string.isRequired
  };

  //maually update the cache so that it matches the data on the server. and updating cache will update the UI
  updateCache = (cache, payload) => {
    //read the cache for the items. we need to do this wit the query used to fetch the items in the first place
    const data = cache.readQuery({ query: ALL_ITEMS_QUERY });
    // remove the deleted ID from cache
    data.items = data.items.filter(item => item.id !== payload.data.deleteItem.id);
    //put the items back
    cache.writeQuery({query: ALL_ITEMS_QUERY, data});
  }

  render() {
    return (
      <Mutation
        mutation={DELETE_ITEM_MUTATION}
        variables={{
          id: this.props.id
        }}
        update={this.updateCache}>
        {(deleteItem, {error, loading}) => (
          <button
          disabled={loading}
          onClick={() => {
            if (confirm('Are you sure you want to delete this item?')){
              deleteItem().catch(err => {
                alert(err.message);
              });
            }
          }}>
            Delete This Item
          </button>
        )}
      </Mutation>
    );
  }
}



export default DeleteItem;