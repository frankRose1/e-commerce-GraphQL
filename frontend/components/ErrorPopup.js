import React from 'react';
import PropTypes from 'prop-types';
import ErrPopUpStyles from './styles/ErrPopUpStyles';

//meant to handle errors on buttons where the UI is too small to display an error message

const ErrorPopup = ({error}) => {
  
  console.log(error);
  return (
    <ErrPopUpStyles showPopup={error ? true : false}>
      <p>{error}</p>
    </ErrPopUpStyles>
  );
};

ErrorPopup.propTypes = {
  error: PropTypes.string.isRequired,
};

export default ErrorPopup;