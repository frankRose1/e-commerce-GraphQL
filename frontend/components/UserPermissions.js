import React, { Component } from 'react';
import { Mutation } from 'react-apollo';
import gql from 'graphql-tag';
import PropTypes from 'prop-types';
import Button from './styles/SickButton';
import Error from './ErrorMessage';

const UPDATE_PERMISSIONS_MUTATION = gql`
  mutation UPDATE_PERMISSIONS_MUTATION(
    $userId: ID!
    $permissions: [Permission]
  ) {
    updateUserPermissions(userId: $userId, permissions: $permissions) {
      id
      name
      email
      permissions
    }
  }
`;

class UserPermissions extends Component {
  static propTypes = {
    user: PropTypes.shape({
      name: PropTypes.string.isRequired,
      email: PropTypes.string.isRequired,
      id: PropTypes.string.isRequired,
      permissions: PropTypes.array.isRequired
    }).isRequired,
    possiblePermissions: PropTypes.array.isRequired
  };

  state = {
    permissions: this.props.user.permissions
  };

  handleClick = async (e, mutation) => {
    const res = await mutation({
      variables: {
        userId: this.props.user.id,
        permissions: this.state.permissions
      }
    });
  };

  handleInputChange = e => {
    const { checked, value } = e.target;
    let updatedPermissions = [...this.state.permissions];
    //figure our if we need to add this permission
    //if its checked, then add it, else remove it
    if (checked) {
      updatedPermissions.push(value);
    } else {
      updatedPermissions = updatedPermissions.filter(
        permission => permission !== value
      );
    }

    this.setState({ permissions: updatedPermissions });
  };

  render() {
    const { user, possiblePermissions } = this.props;
    const { permissions } = this.state;
    const { email, name, id } = user;

    return (
      <Mutation mutation={UPDATE_PERMISSIONS_MUTATION}>
        {(updateUserPermissions, { loading, error }) => (
          <>
            {error && (
              <tr>
                <td colspan='8'>
                  <Error error={error} />
                </td>
              </tr>
            )}
            <tr data-test='permissions-row'>
              <td>{name}</td>
              <td>{email}</td>
              {possiblePermissions.map((permission, i) => (
                <td key={i}>
                  <label htmlFor={`${id}-permission-${permission}`}>
                    <input
                      type='checkbox'
                      id={`${id}-permission-${permission}`}
                      checked={permissions.includes(permission)}
                      value={permission}
                      disabled={loading}
                      onChange={this.handleInputChange}
                    />
                  </label>
                </td>
              ))}
              <td>
                <Button
                  type='button'
                  data-test='update-button'
                  disabled={loading}
                  onClick={e => this.handleClick(e, updateUserPermissions)}
                >
                  Updat{loading ? 'ing' : 'e'}
                </Button>
              </td>
            </tr>
          </>
        )}
      </Mutation>
    );
  }
}

export default UserPermissions;
export { UPDATE_PERMISSIONS_MUTATION };
