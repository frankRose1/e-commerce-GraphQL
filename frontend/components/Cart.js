import React from 'react';
import {Mutation, Query} from 'react-apollo';
import gql from 'graphql-tag';
import {adopt} from 'react-adopt';
import User from './User';
import CartItem from './CartItem';
import CartStyles from './styles/CartStyles';
import Supreme from './styles/Supreme';
import CloseButton from './styles/CloseButton';
import Button from './styles/SickButton';
import PurchaseItem from './PurchaseItem';
import calcTotalPrice from '../lib/calcTotalPrice';
import formatMoney from '../lib/formatMoney';

//Client side query & mutation for cart boolean
const LOCAL_STATE_QUERY = gql`
  query {
    cartOpen @client
  }
`;

const TOGGLE_CART_MUTATION = gql`
  mutation {
    toggleCart @client
  }
`;

const Composed = adopt({
  user: ({render}) => <User>{render}</User>,
  toggleCart: ({render}) => <Mutation mutation={TOGGLE_CART_MUTATION}>{render}</Mutation>,
  localState: ({render}) => <Query query={LOCAL_STATE_QUERY}>{render}</Query>
});

const Cart = () => {
  return (
    <Composed>
      {( {user, toggleCart, localState} ) => {
        const {data: {me}, loading} = user;
        if (loading) return <p>Loading...</p>
        if(!me) return null //dont show the cart if user is not signed in
        return (
          <CartStyles
            data-test="cart"
            open={localState.data.cartOpen}>
            <header>
              <CloseButton onClick={toggleCart} title="close">&times;</CloseButton>
              <Supreme>{me.name}'s Cart</Supreme>
              <p>{me.cart.length} item{me.cart.length === 1 ? '' : 's'} in your cart</p>
            </header>

              <ul>
                {me.cart.map(cartItem => (
                  <CartItem 
                    key={cartItem.id}
                    cartItem={cartItem}/>
                ))}
              </ul>

            <footer>
              <p>{formatMoney(calcTotalPrice(me.cart))}</p>
              {me.cart.length
                ? (
                  <PurchaseItem>
                    <Button>Checkout</Button>
                  </PurchaseItem>
                )
              : null }
            </footer>
          </CartStyles>
        )
      }}
    </Composed>
  );
};

export default Cart;
export {TOGGLE_CART_MUTATION, LOCAL_STATE_QUERY};