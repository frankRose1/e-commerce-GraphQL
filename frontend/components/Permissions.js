import React from 'react';
import {Query} from 'react-apollo';
import gql from 'graphql-tag';
import Error from './ErrorMessage';
import Table from './styles/Table';
import UserPermissions from './UserPermissions';

const possiblePermissions = [
  'ADMIN',
  'USER',
  'ITEMCREATE',
  'ITEMUPDATE',
  'ITEMDELETE',
  'PERMISSIONUPDATE'
];

const ALL_USERS_QUERY = gql`
  query ALL_USERS_QUERY {
    usersPermissions {
      id
      name
      email
      permissions
    }
  }
`;

const Permissions = () => (
  <Query query={ALL_USERS_QUERY}>
    {({data, loading, error}) => (
      <div>
        <Error error={error}/>
        <div>
          <h2>Manage Permissions</h2>
          <Table>
            <thead>
              <tr>
                <th>Name</th>
                <th>Email</th>
                {possiblePermissions.map((perm, i) => (
                  <th key={i}>{perm}</th>
                ))}
                <th>Update</th>
              </tr>
            </thead>
            <tbody>
              {data.usersPermissions.map(user => (
                <UserPermissions
                  key={user.id}
                  possiblePermissions={possiblePermissions}
                  user={user}/>
              ))}
            </tbody>
          </Table>
        </div>
      </div>
    )}
  </Query>
);

export default Permissions;
export {ALL_USERS_QUERY};