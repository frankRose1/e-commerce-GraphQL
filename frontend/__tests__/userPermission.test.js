import { mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import { MockedProvider } from 'react-apollo/test-utils';
import wait from 'waait';
import { fakeUser } from '../lib/testUtils';
import UserPermissions, {
  UPDATE_PERMISSIONS_MUTATION
} from '../components/UserPermissions';
import { possiblePermissions } from '../components/Permissions';

const user = {
  ...fakeUser(),
  permissions: ['ADMIN', 'USER']
};

const mocks = [
  {
    request: {
      query: UPDATE_PERMISSIONS_MUTATION,
      variables: {
        userId: user.id,
        permissions: user.permissions
      }
    },
    result: {
      data: {
        updateUserPermissions: { ...fakeUser() }
      }
    }
  }
];

describe('<UserPermissions />', () => {
  it('renders and matches the snapsot', async () => {
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <UserPermissions
          user={user}
          possiblePermissions={possiblePermissions}
        />
      </MockedProvider>
    );
    await wait();
    wrapper.update();
    const row = wrapper.find('tr[data-test="permissions-row"]');
    expect(toJSON(row)).toMatchSnapshot();
  });

  it('checks off the inputs by default if the user has a certain permission', () => {
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <table>
          <tbody>
            <UserPermissions
              user={user}
              possiblePermissions={possiblePermissions}
            />
          </tbody>
        </table>
      </MockedProvider>
    );
    const row = wrapper.find('tr[data-test="permissions-row"]');
    expect(row.find(`[value="${user.permissions[0]}"]`).props().checked).toBe(
      true
    );
    expect(row.find(`[value="${user.permissions[1]}"]`).props().checked).toBe(
      true
    );
  });

  it('updates state on input change', () => {
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <table>
          <tbody>
            <UserPermissions
              user={user}
              possiblePermissions={possiblePermissions}
            />
          </tbody>
        </table>
      </MockedProvider>
    );
    const row = wrapper.find('tr[data-test="permissions-row"]');
    //check itemcreate and check itemdelete
    row
      .find('input[value="ITEMCREATE"]')
      .simulate('change', { target: { checked: true, value: 'ITEMCREATE' } });
    row
      .find('input[value="ITEMDELETE"]')
      .simulate('change', { target: { checked: true, value: 'ITEMDELETE' } });
    const component = wrapper.find('UserPermissions').instance();
    expect(component.state).toMatchObject({
      permissions: ['ADMIN', 'USER', 'ITEMCREATE', 'ITEMDELETE']
    });
  });

  it('properly calls the mutation', async () => {
    const updateUserPermissionsMock = jest.fn().mockResolvedValue({
      data: {
        updateUserPermissions: { ...fakeUser() }
      }
    });
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <table>
          <tbody>
            <UserPermissions
              user={user}
              possiblePermissions={possiblePermissions}
            />
          </tbody>
        </table>
      </MockedProvider>
    );
    const component = wrapper.find('UserPermissions').instance();
    component.handleClick({}, updateUserPermissionsMock);
    expect(updateUserPermissionsMock).toHaveBeenCalledWith({
      variables: {
        userId: user.id,
        permissions: ['ADMIN', 'USER']
      }
    });
  });
});
