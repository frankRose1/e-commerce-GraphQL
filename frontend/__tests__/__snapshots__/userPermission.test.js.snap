// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UserPermissions /> renders and matches the snapsot 1`] = `
<tr
  data-test="permissions-row"
>
  <td>
    Miss Coleman Berge
  </td>
  <td>
    Delmer.Smith@yahoo.com
  </td>
  <td
    key="0"
  >
    <label
      htmlFor="4234-permission-ADMIN"
    >
      <input
        checked={true}
        disabled={false}
        id="4234-permission-ADMIN"
        onChange={[Function]}
        type="checkbox"
        value="ADMIN"
      />
    </label>
  </td>
  <td
    key="1"
  >
    <label
      htmlFor="4234-permission-USER"
    >
      <input
        checked={true}
        disabled={false}
        id="4234-permission-USER"
        onChange={[Function]}
        type="checkbox"
        value="USER"
      />
    </label>
  </td>
  <td
    key="2"
  >
    <label
      htmlFor="4234-permission-ITEMCREATE"
    >
      <input
        checked={false}
        disabled={false}
        id="4234-permission-ITEMCREATE"
        onChange={[Function]}
        type="checkbox"
        value="ITEMCREATE"
      />
    </label>
  </td>
  <td
    key="3"
  >
    <label
      htmlFor="4234-permission-ITEMUPDATE"
    >
      <input
        checked={false}
        disabled={false}
        id="4234-permission-ITEMUPDATE"
        onChange={[Function]}
        type="checkbox"
        value="ITEMUPDATE"
      />
    </label>
  </td>
  <td
    key="4"
  >
    <label
      htmlFor="4234-permission-ITEMDELETE"
    >
      <input
        checked={false}
        disabled={false}
        id="4234-permission-ITEMDELETE"
        onChange={[Function]}
        type="checkbox"
        value="ITEMDELETE"
      />
    </label>
  </td>
  <td
    key="5"
  >
    <label
      htmlFor="4234-permission-PERMISSIONUPDATE"
    >
      <input
        checked={false}
        disabled={false}
        id="4234-permission-PERMISSIONUPDATE"
        onChange={[Function]}
        type="checkbox"
        value="PERMISSIONUPDATE"
      />
    </label>
  </td>
  <td>
    <SickButton
      data-test="update-button"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      <button
        className="SickButton-epeo8a-0 hDvtOo"
        data-test="update-button"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Updat
        e
      </button>
    </SickButton>
  </td>
</tr>
`;
